### PR TITLE
Add container storage metric

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
For enablement of KEP-4191, we realize that having a metric would be useful. This metric lives in the container runtime because container runtime configures the storage.

We add container_storage as a metric and label it 0 or 1 for the type of storage we have.
#### Does this PR introduce a user-facing change?

```release-note
Add the metric storage_configuration_type to distinguish between image_storage usage and default usage.
```

### CODE REVIEW

File: server/metrics/metrics.go [hunk 1]
```diff
@@ -401,6 +410,19 @@ func (m *Metrics) MetricResourcesStalledAtStage(stage string) {
 	c.Inc()
 }
 
+func (m *Metrics) MetricStorageContainerType(imageStore string) {
+	c, err := m.metricContainerStorageType.GetMetricWithLabelValues("container_storage")
+	if err != nil {
+		logrus.Warnf("Unable to write container storage metric: %v", err)
+	}
+	var typeOfStorage float64 = 0
```
Comment (haircommander) [inline server/metrics/metrics.go:418]: no way to set separate filesystem now?

File: server/server.go [hunk 2]
```diff
@@ -562,6 +562,9 @@ func New(
 		return nil, err
 	}
 
+	// Set container storage metrics
+	metrics.Instance().MetricStorageContainerType(s.getInfo().StorageImage)
```
Comment (haircommander) [inline server/server.go:566]: can you add a comment on why this is here?

File: server/metrics/metrics.go (hunk 1 above)
Comment (kannon92) [reply server/metrics/metrics.go:418]: There is. Do you know if we expose the constant we use for graphroot anywhere?

Comment (kannon92) [reply server/metrics/metrics.go:418]: I didn't add it because we aren't technically supporting this in any feature at the moment.

Comment (haircommander) [reply server/metrics/metrics.go:418]: in
```
vendor/github.com/containers/storage/types/options_linux.go:	defaultGraphRoot string = "/var/lib/containers/storage"

```
so not exposed. the other thing to think about is we don't know if it's a split filesystem if we don't check the disk is  separated...

Comment (kannon92) [reply server/metrics/metrics.go:418]: Yea but imo from the crio side, I don't think that is that important of an distinction.

We wouldn't behave any differently if /var/lib/images was on same disk or not. I think the logic I have for imageStore is okay.

Comment (haircommander) [reply server/metrics/metrics.go:418]: yeah I mostly mean for split fs when graphroot is on a separate disk

File: server/metrics/metrics.go [hunk 3]
```diff
@@ -239,6 +241,14 @@ func New(config *libconfig.MetricsConfig) *Metrics {
 			},
 			[]string{"stage"},
 		),
+		metricContainerStorageType: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Subsystem: collectors.Subsystem,
+				Name:      collectors.StorageConfigurationType.String(),
+				Help:      "Type of container storage crio is configured with. 0 is default. 1 is split filesystem. 2 is separate filesystem.",
```
Comment (kwilczynski) [inline server/metrics/metrics.go:248]: It's CRI-O.

```suggestion
				Help:      "Type of container storage CRI-O is configured with. 0 is default. 1 is split filesystem. 2 is separate filesystem.",
```

File: docs/crio.8.md [hunk 4]
```diff
@@ -386,9 +386,9 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
 
 **--read-only**: Setup all unprivileged containers to run as read-only. Automatically mounts the containers' tmpfs on '/run', '/tmp' and '/var/tmp'.
 
-**--root, -r**="": The CRI-O root directory. (default: "/var/lib/containers/storage")
+**--root, -r**="": The CRI-O root directory. (default: "/home/kehannon/.local/share/containers/storage")
```
Comment (kwilczynski) [inline docs/crio.8.md:389]: Needs to be manually adjusted.

File: server/otel-collector/collectors/collectors.go [hunk 5]
```diff
@@ -63,6 +63,9 @@ const (
 
 	// ResourcesStalledAtStage is the key for the resources stalled at different stages in container and pod creation.
 	ResourcesStalledAtStage Collector = crioPrefix + "resources_stalled_at_stage"
+
+	// StorageConfigurationType is the key for the type of storage cri-o is configured with.
```
Comment (kwilczynski) [inline server/otel-collector/collectors/collectors.go:67]:
```suggestion
	// StorageConfigurationType is the key for the type of storage CRI-O is configured with.
```

File: server/server.go [hunk 6]
```diff
@@ -562,6 +562,10 @@ func New(
 		return nil, err
 	}
 
+	// Set container storage metrics
```
Comment (kwilczynski) [inline server/server.go:565]:
```suggestion
	// Set container storage metrics.
```

File: docs/crio.8.md [hunk 7]
```diff
@@ -386,9 +386,9 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
...
 
-**--runroot**="": The CRI-O state directory. (default: "/run/containers/storage")
+**--runroot**="": The CRI-O state directory. (default: "/run/user/4213954/containers")
```
Comment (kwilczynski) [inline docs/crio.8.md:391]: Ditto as above.

File: docs/crio.8.md [hunk 8]
```diff
@@ -332,7 +332,7 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
 
 **--metrics-cert**="": Certificate for the secure metrics endpoint.
 
-**--metrics-collectors**="": Enabled metrics collectors. (default: "image_pulls_layer_size", "containers_events_dropped_total", "containers_oom_total", "processes_defunct", "operations_total", "operations_latency_seconds", "operations_latency_seconds_total", "operations_errors_total", "image_pulls_bytes_total", "image_pulls_skipped_bytes_total", "image_pulls_failure_total", "image_pulls_success_total", "image_layer_reuse_total", "containers_oom_count_total", "containers_seccomp_notifier_count_total", "resources_stalled_at_stage")
+**--metrics-collectors**="": Enabled metrics collectors. (default: "image_pulls_layer_size", "containers_events_dropped_total", "containers_oom_total", "processes_defunct", "operations_total", "operations_latency_seconds", "operations_latency_seconds_total", "operations_errors_total", "image_pulls_bytes_total", "image_pulls_skipped_bytes_total", "image_pulls_failure_total", "image_pulls_success_total", "image_layer_reuse_total", "containers_oom_count_total", "containers_seccomp_notifier_count_total", "resources_stalled_at_stage", "storage_configuration_type")
```
Comment (kwilczynski) [inline docs/crio.8.md:335]: Which metric is it? Change description says this:

> Add the metric container_storage to distinguish between image_storage usage and default usage.

Comment (kannon92) [reply docs/crio.8.md:335]: updated the change description.

File: server/metrics/metrics.go (hunk 3 above)
Comment (kannon92) [reply server/metrics/metrics.go]: sorry about that. updated.

File: docs/crio.8.md (hunk 4 above)
Comment (kannon92) [reply docs/crio.8.md]: yikes. good find.

File: server/otel-collector/collectors/collectors.go (hunk 5 above)
Comment (kannon92) [reply server/otel-collector/collectors/collectors.go]: fixed.

File: server/server.go (hunk 6 above)
Comment (kannon92) [reply server/server.go]: updated.

File: docs/crio.8.md (hunk 7 above)
Comment (kannon92) [reply docs/crio.8.md]: updated.

**Comment by kannon92** [comment]
/hold

We will continue to discuss this feature so not sure if this should merge yet.

File: docs/crio.8.md (hunk 8 above)
Comment (saschagrunert) [inline docs/crio.8.md:335]: The metric needs to be added to https://github.com/cri-o/cri-o/blob/main/tutorials/metrics.md

**Comment by kannon92** [comment]
/close

I don't think we need this yet. Still figuring out design plan for split filesystem

**Pull request closed without merging**